contributors/design-proposals/pod-resource-management.md

```diff
@@ -1,114 +1,419 @@
+
+All writes to the QoS level cgroup sandboxes are protected via a
+common lock in the kubelet to ensure we do not have multiple concurrent
+writers to this tier in the hierarchy.
+
+### QoS level CPU allocation
+
+The `BestEffort` cgroup sandbox is statically configured as follows:
+
+```
+ROOT/BestEffort/cpu.shares = 2
+```
+
+This ensures that allocation of CPU time to pods in this QoS class
+is given the lowest priority.
+
+The `Burstable` cgroup sandbox CPU share allocation is dynamic based
+on the set of pods currently scheduled to the node.  
+
+```
+ROOT/Burstable/cpu.shares = max(sum(Burstable pods cpu requests, 2)
+```
+
+The Burstable cgroup sandbox is updated dynamically in the exit
+points described in the previous section.  Given the compressible
+nature of CPU, and the fact that cpu.shares are evaluated via relative
+priority, the risk of an update being incorrect is minimized as the `kubelet`
+converges to a desired state.  Failure to set `cpu.shares` at the QoS level
+cgroup would result in `500m` of cpu for a Guaranteed pod to have different
+meaning than `500m` of cpu for a Burstable pod in the current hierarchy.
+For this reason, we will always set `cpu.shares` for the QoS level sandboxes
+by default as part of roll-out for this feature.
+
+### QoS level memory allocation
+
+By default, no memory limits are applied to the BestEffort
+and Burstable QoS level cgroups unless a `--qos-reserve-requests` value
+is specified for memory.
+
+The heuristic that is applied is as follows for each QoS level sandbox:
+
+```
+ROOT/Burstable/memory.limit_in_bytes = 
+    Node.Allocatable - {(summation of memory requests of `Guaranteed` pods)*(reservePercent / 100)}
+ROOT/BestEffort/memory.limit_in_bytes = 
+    Node.Allocatable - {(summation of memory requests of all `Guaranteed` and `Burstable` pods)*(reservePercent / 100)}
+```
+
+A value of `--experimental-qos-reserve-requests=memory=100%` will cause the
+`kubelet` to adjust the Burstable and BestEffort cgroups from consuming memory
+that was requested by a higher QoS class. This increases the risk
+of inducing OOM on BestEffort workloads in favor of increasing memory
+resource guarantees for Burstable and Guaranteed workloads.  A value of 
+`--experimental-qos-reserve-requests=memory=0%` will allow a Burstable
+and BestEffort QoS sandbox to consume up to the full allocatable amount if
+available in favor of increasing memory resource guarantees for Guaranteed
+workloads.
+
+Since memory is an incompressible resource, it is possible that a QoS
+level cgroup sandbox may not be able to reduce memory usage below the
+value specified in the heuristic during pod admission and pod termination.
+As a result, the `kubelet` runs a periodic thread to attempt to converge
+to this desired state from the above heuristic.  If unreclaimable memory
+usage has exceeded the desired limit for the sandbox, the `kubelet` will
+attempt to set the effective limit near the current usage to put pressure
```

Comment:
@vishh -- does something like the following help?  I disagree on the safety argument.  Operators most concerned about safety may deploy G workloads first to a node before opening up other workloads for scheduling.  For nodes with fluid set of pods being scheduled, I want to attempt to provide safety, but do no worse than if the reservation feature is disabled.

**Use cases:**
1. I want to prioritize access to compressible resources for my system/cluster daemons over end-user pods.
2. I want to prioritize access to compressible resources for my G workloads over my Bu workloads.
3. I want to prioritize access to compressible resources for my Bu workloads over my Be workloads.

An operator may choose to prioritize access to a compressible resource over utilization for each step in the hierarchy of workloads described above dependent upon their intended workload.  Almost all cluster operators will use node allocatable to enforce the first use case in order to provide reliable operation of the node.  It is understood that not all operators may feel the same for G/Bu workloads across the user community.  For the users I represent, I know they may opt to deploy important cluster services as G workloads via a `DaemonSet` and would like a similar resource reservation model as is provided via node allocatable for system/cluster daemons and end-user pods.  In addition, I know users I represent would like to run Be workloads and minimize their ability to impact Bu pods.  Reliance on eviction, oom_killer, and critical pods is not always sufficient.  Eviction still has latency, oom_killer is disruptive, and critical pods as I view them are intended to provide G style guarantees but run with Bu resource requirements.

**Design Principle**
kubelet with opt-in configuration will _attempt_ to limit the ability for a pod in a lower QoS tier to burst utilization of a compressible resource that was requested by a pod in a higher QoS tier.

**Mechanism**
Prior to starting a G pod, we will always attempt to update the Bu and Be tiers to have their memory limits reduced based on the incoming G pod memory request.  It is possible that we are unable to reduce the Bu and Be tier to their new desired limit if usage already has exceeded that value prior to execution of the G pod.  If they have exceeded their usage, we will set the limit at or near their usage to put pressure on the cgroup to prevent further growth.  The kubelet will not wait for the QoS cgroup memory limit to converge to the desired state prior to execution of the G pod.  This does mean that the G pod could induce an OOM for the kubepods cgroup, but per our QoS design, we would prefer the oom_killer targets the Bu/Be pods first, or ideally, we get the eviction code path to kick in response to kernel notification, and evict the Bu/Be pod.  The periodic task that attempts to converge the QoS tier memory limits would then have a better opportunity to converge on the desired state so any future Bu/Be pods that land do not impact the G pod since they would be working under a reduced memory limit.  The same logic applies for the Bu pods that get scheduled, but reduces the pertinent QoS cgroup sandbox update to just Be tier.

**Best Practices**
Operators that want to provide a similar resource reservation model for G pods as we offer via enforcement of node allocatable are encouraged to schedule their G pods via a DameonSet prior to opening up the node for scheduling of Bu/Be workloads.  This ensures that the Bu/Be tiers have had their QoS memory limits appropriately adjusted before taking unbounded workloads on the node.

**Future considerations**
It's possible some users may want to stall execution of a pod in a higher QoS tier if resource reservations for compressible resources have not converged on the desired state.  It can be evaluated based on more experience in the community.

We have discussed potentially allow end-user pods to schedule in the kube/system cgroup spaces, but the same hierarchy of needs may still be extended.  For example, hosted multi-tenant shared-node offerings may want to allow the operators to deploy pods to kube/system cgroup spaces, but not the tenants on a shared node.  For those users, offering the same opportunity to reserve requests at each level in the hierarchy is important.